File: BleAntilost/ble-anti-lost-ios/bleAntiLost/bleAntiLost/QRCodeScannerViewController.m
//
//  QRCodeScannerViewController.m
//  bleAntiLost
//
//  Created by china topflytech on 2024/6/14.
//
// QRCodeScannerViewController.m
// QRCodeScannerViewController.m
#import "QRCodeScannerViewController.h"

@interface QRCodeScannerViewController ()
@property (nonatomic, strong) AVCaptureSession *captureSession;
@property (nonatomic, strong) AVCaptureVideoPreviewLayer *videoPreviewLayer;
@end

@implementation QRCodeScannerViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor blackColor];
    [self setupCaptureSession];
    [self startCaptureSession];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    [self.captureSession stopRunning];
}

- (void)setupCaptureSession {
    NSError *error;
    self.captureSession = [[AVCaptureSession alloc] init];
    
    // 设置输入设备
    AVCaptureDevice *videoCaptureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
    AVCaptureDeviceInput *videoInput = [AVCaptureDeviceInput deviceInputWithDevice:videoCaptureDevice error:&error];
    
    if (videoInput) {
        [self.captureSession addInput:videoInput];
    } else {
        NSLog(@"Error: %@", error);
        return;
    }
    
    // 设置输出设备
    AVCaptureMetadataOutput *metadataOutput = [[AVCaptureMetadataOutput alloc] init];
    [self.captureSession addOutput:metadataOutput];
    
    [metadataOutput setMetadataObjectsDelegate:self queue:dispatch_get_main_queue()];
    [metadataOutput setMetadataObjectTypes:@[AVMetadataObjectTypeQRCode]];
    
    // 添加视频预览层
    self.videoPreviewLayer = [AVCaptureVideoPreviewLayer layerWithSession:self.captureSession];
    self.videoPreviewLayer.frame = self.view.layer.bounds;
    self.videoPreviewLayer.videoGravity = AVLayerVideoGravityResizeAspectFill;
    [self.view.layer addSublayer:self.videoPreviewLayer];
}

- (void)startCaptureSession {
    // 使用后台队列启动 AVCaptureSession
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        [self.captureSession startRunning];
    });
}

// 处理扫描到的二维码
- (void)captureOutput:(AVCaptureOutput *)output didOutputMetadataObjects:(NSArray<__kindof AVMetadataObject *> *)metadataObjects fromConnection:(AVCaptureConnection *)connection {
    if (metadataObjects.count == 0) {
        return;
    }
    
    AVMetadataMachineReadableCodeObject *metadataObject = [metadataObjects objectAtIndex:0];
    if ([metadataObject.type isEqualToString:AVMetadataObjectTypeQRCode]) {
        NSString *qrCode = metadataObject.stringValue;
        NSLog(@"Scanned QR Code: %@", qrCode);
        
        // 停止扫描
        [self.captureSession stopRunning];
        
        // 使用主线程调用代理方法
        dispatch_async(dispatch_get_main_queue(), ^{
            if ([self.delegate respondsToSelector:@selector(didFindQRCode:)]) {
                [self.delegate didFindQRCode:qrCode];
            }
            [self dismissViewControllerAnimated:YES completion:nil];
        });
    }
}

@end

File: IOS-TFTBLE/tftble/QRCodeScannerViewController.swift
//
//  QRCodeScannerViewController.swift
//  tftble
//
//  Created by china topflytech on 2023/12/21.
//  Copyright © 2023 com.tftiot. All rights reserved.
//

import AVFoundation
import UIKit

class QRCodeScannerViewController: UIViewController, AVCaptureMetadataOutputObjectsDelegate {
    var captureSession: AVCaptureSession!
    var videoPreviewLayer: AVCaptureVideoPreviewLayer!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        let titleLabel = UILabel(frame: CGRect(x: 0, y: 0, width: 40, height: 40))
        titleLabel.text = NSLocalizedString("qrcode_scan", comment: "QR code scanning")
        self.navigationItem.titleView = titleLabel
        // 创建 AVCaptureSession 实例
        captureSession = AVCaptureSession()
        
        // 获取摄像头设备
        guard let videoCaptureDevice = AVCaptureDevice.default(for: .video) else { return }
        
        // 创建 AVCaptureDeviceInput 实例
        guard let videoInput = try? AVCaptureDeviceInput(device: videoCaptureDevice) else { return }
        
        // 将输入设备添加到会话
        if captureSession.canAddInput(videoInput) {
            captureSession.addInput(videoInput)
        } else {
            return
        }
        
        // 创建 AVCaptureMetadataOutput 实例
        let metadataOutput = AVCaptureMetadataOutput()
        
        // 将输出添加到会话
        if captureSession.canAddOutput(metadataOutput) {
            captureSession.addOutput(metadataOutput)
            
            // 设置委托和扫描类型
            metadataOutput.setMetadataObjectsDelegate(self, queue: DispatchQueue.main)
            metadataOutput.metadataObjectTypes = [.qr]
        } else {
            return
        }
        
        // 创建预览图层
        videoPreviewLayer = AVCaptureVideoPreviewLayer(session: captureSession)
        videoPreviewLayer.videoGravity = .resizeAspectFill
        videoPreviewLayer.frame = view.layer.bounds
        view.layer.addSublayer(videoPreviewLayer)
        
        // 启动会话
        captureSession.startRunning()
    }
    var delegate : QrCodeScanDelegate?
    // 代理方法，处理扫描结果
    func metadataOutput(_ output: AVCaptureMetadataOutput, didOutput metadataObjects: [AVMetadataObject], from connection: AVCaptureConnection) {
        guard let metadataObject = metadataObjects.first as? AVMetadataMachineReadableCodeObject,
              let qrCodeString = metadataObject.stringValue else {
            return
        }
        
        // 处理扫描到的二维码字符串
        print("扫描到的二维码：\(qrCodeString)")
        
        // 停止会话
        captureSession.stopRunning()
        self.delegate?.setQrcodeValue(value: qrCodeString)
        self.navigationController?.popViewController(animated: false)
    }
}
